### DISCUSSION TITLE
How to run actions by Pipelines, not by jobs

### DISCUSSION THREAD

**Discussion by narmatov-asylbek** [community/community#25258] [Actions]
<p>Hey, everyone!</p>
<p>I am hosting my own runner for Github Actions. And I have one problem: I need to make actions run by pipelines, not by jobs.</p>
<p>Suppose, I have a testing.yml that has two jobs: testing and cleanup-containers that depend on testing. And the actions will trigger on the pull request.</p>
<p>If two people create pull request jobs will run in the following order: testing 1, testing 2, cleanup-containers 1, cleanup-containers 2.</p>
<p>I need to make actions run in this order: testing 1, cleanup-containers 1, testing 2, cleanup-containers 2.</p>
<p>If you have any suggestions, please, let me know.</p>

**Comment by airtower-luna** [comment]
<p>If steps need to run on the same runner and without anything in between put them into the same job. Each job is supposed to run completely separate from any other (other than <code>needs</code> relationships and artifact transfer, if any).</p>
<p>From the job names I guess that the “cleanup-containers” job is supposed to delete containers created by the “testing” job. Just move the cleanup to the end of the testing job (basically make it clean up after itself). You might want to use <code>if: always()</code> or something similar so the cleanup runs even if the tests failed.</p>

**Comment by jsoref** [comment, marked as answer]
<p>If you’re using self-hosted runners, you could use a job to define a <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix">matrix</a> that picks the self-hosted runner for all of the remaining jobs.</p>
<p>The rest of it could be accomplished with a lot of <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds"><code>needs:</code></a>.</p>
